src/kio/serial/readers.py

```diff
@@ -23,46 +23,54 @@
 from kio.static.primitive import u32
 from kio.static.primitive import u64
 
+from .errors import BufferUnderflow
 from .errors import UnexpectedNull
 
 T = TypeVar("T")
 Reader: TypeAlias = Callable[[IO[bytes]], T]
 
 
+def read_exact(buffer: IO[bytes], num_bytes: int) -> bytes:
+    value = buffer.read(num_bytes)
```

Comment:
Read can return `None` if the object is `io.RawIOBase`. If this is always expected to be buffered maybe the type should be more exact?

Reply:
> Read can return `None` if the object is `io.RawIOBase`.

This means `RawIOBase` is not a subtype of `typing.IO`, so I think the typing here is fine?

https://github.com/python/typeshed/blob/faf9d77d79adcda3bd6f8f170c7f76e30080c1d2/stdlib/typing.pyi#L722

This gives a type error:

```python
import io
from typing import IO
def takes_io(b: IO[bytes]) -> None:
    ...
takes_io(io.RawIOBase())
```

```
prog.py:5: error: Argument 1 to "takes_io" has incompatible type "RawIOBase"; expected "IO[bytes]"  [arg-type]
    takes_io(io.RawIOBase())
             ^~~~~~~~~~~~~~
```

---

On this level I don't think there's any set expectation for whether underlying object should be buffered or not, anything satisfying `IO[bytes]` is OK, however I made a choice to cater for reading full responses into memory before parsing them. Instead of parsing them directly from the network stream.

Initially this was not the case, and it meant all readers had to be implemented as generators to support multiple kinds of sources for bytes. This is where I decided to scratch that idea: https://github.com/Aiven-Open/kio/pull/44/.

Perhaps the parameter naming gives some confusion here, and we should consider not using `buffer`?